store/src/state.rs

```diff
@@ -220,13 +413,39 @@ impl State {
 // UTILITIES
 // ================================================================================================
 
+/// Returns the nullifier's block number given its leaf value in the TSMT.
+fn block_to_nullifier_data(block: BlockNumber) -> Word {
+    [Felt::new(block as u64), Felt::ZERO, Felt::ZERO, Felt::ZERO]
+}
+
+// /// Returns the block number encoded as a leaf value to be used in the TSMT.
+// fn nullifier_data_to_block(block: Word) -> BlockNumber {
+//     block[0].as_int() as BlockNumber
+// }
+
+/// Creates a [SimpleSmt] tree from the `notes`.
+pub fn build_notes_tree(notes: &[Note]) -> Result<SimpleSmt, anyhow::Error> {
+    let mut entries: Vec<(u64, Word)> = Vec::with_capacity(notes.len() * 2);
+
+    for (index, note) in notes.iter().enumerate() {
+        let note_hash = note.note_hash.clone().ok_or(StateError::MissingNoteHash)?;
+        let account_id = note.sender.try_into().or(Err(StateError::InvalidAccountId))?;
+        let note_metadata = NoteMetadata::new(account_id, note.tag.into(), note.num_assets.into());
+        let index = (index as u64) * 2;
+        entries.push((index, note_hash.try_into()?));
+        entries.push((index + 1, note_metadata.into()));
+    }
```

Comment:
I have thought about that. Not that easy. The loop above has a protobuf digest, SimpleSMT only work with RpoDigest, the conversion among the two is fallible (because it is used in the request handler to validate network data, see https://github.com/0xPolygonMiden/miden-node/issues/13).

That means the iterator returns a `Iterator<Item=Result<(u64, RpoDigest),_>>` and we need it to be `Iterator<Item=(u64, RpoDigest)>`. To do this conversion the iterator has to be consumed, so we have the vector.

To fix that we need to either change the data in the DB, so very write/read would pay the conversion. Or make the `SimpleSMT` generic over its value type, so that protobuf's version of the digest can be used.

Reply:
Let's create an issue to discuss these options further.